Rando: Add skullsanity and nighttime GS tokens require Sun's Song

aMannus opened the pull request:
This _should_ add all 100 tokens correctly so they can be shuffled with the Tokensanity option.

I haven't tested every single possible item that can show up in them yet, but the ones I did test worked. I tested some random equipment items and a consumable, but this might need some more testing to make sure.

Also, I counted the added skulls and it does add up to 100, but I'd like to have someone else test all 100 once this is merged. I have a spoiler log ready where every spider is changed to anything but a skull token for that. 

It also adds a custom text box for when you pick up a skull token that got shuffled to an ice trap. In scenarios where the player is not able to receive an ice trap, the game will continue to try and grant the ice trap until link is in his frozen state.

I also added the nighttime skulls requiring Sun's Song option since it makes sense to have that when skullsanity is a thing. This should only be a logic thing, so it was just a matter of exposing it to the 3DS logic. I placed it in a spot in the settings menu I'm not super happy with, but I'm not sure where else to put it since we're still missing quite a lot of the other options.

Any feedback is appreciated, otherwise go ahead and merge it in, and I'll ask some people in the randomizer-playtesting channel to test this and the UX changes.

aMannus committed:
not yet

aMannus committed:
Merge branch 'develop' into spoopy-spiders

aMannus committed:
still nothing yet

aMannus committed:
Closer but not yet

aMannus committed:
Exposed static table to change skulltokens

This static table is normally only used within z_player.c, but is needed to switch the itemId to something Item_Give() can use and to set the correct TextId.

aMannus committed:
Freeze fix/first batch of skull actors

aMannus committed:
Cleanup and freeze fix

Don't freeze the player when the skulltula token check is something else than a token. Also some cleanup and declaring variables more globally within z_en_si.c

aMannus committed:
Fixed mistake

aMannus committed:
Another batch of spooders

aMannus committed:
Added v1 ice traps and fixed some tokens

aMannus committed:
Merge branch 'develop-zhora' into spoopy-spiders

aMannus committed:
Removed temporary bunny hood change

aMannus committed:
Fix after dev merge

aMannus committed:
Fixed all possible ice trap scenarios

aMannus committed:
Added comment

aMannus committed:
Merge branch 'rando-next' into spoopy-spiders

aMannus committed:
Small fix after UX merge

aMannus committed:
Nother batch of spooders

aMannus committed:
Added all remaining tokens(?)

aMannus committed:
Add setting tooltip and fixed an oopsie

aMannus committed:
Added nighttime GS sun's song option

Flax4 commented:
Spawning a new skulltula check in the same area seems to replace both checks with the same item.

https://www.youtube.com/watch?v=Pmo0xFv9Fjc

aMannus committed:
Fix for spooders to overwroote each ooder

aMannus commented:
> Spawning a new skulltula check in the same area seems to replace both checks with the same item.
> 
> https://www.youtube.com/watch?v=Pmo0xFv9Fjc

Just to add this here since it was only discussed in discord: this is fixed now.

briaguya0 commented:
ice traps are giving skull tokens (`Item_Give` is still being called) 

https://user-images.githubusercontent.com/70942617/180073929-0de52fc9-dff8-4338-972c-cb1267473afd.mp4

aMannus committed:
Fixes for double spiders & ice traps

getItemId should've been set earlier to make sure double spiders didn't screw up ice traps.

Additionally Item_Give() was still being called with skultulla tokens as the default value so it granted a token when picking up an ice trap.

Skulltula freeze was also happening on ice traps which it shouldn't do, so that's fixed now too.

aMannus commented:
> ice traps are giving skull tokens (`Item_Give` is still being called)
> 
>  2022-07-20.16-17-58.mp4

Good catch. I also found some other mistakes that had to do with the skulltula freeze when picking up an ice trap without the no freeze enhancement and something else that could've screwed up when picking up an ice trap when multiple skulltula rewards were in-world, so that should also be fixed now.

Flax4 commented:
Ice trap tokens doesn't seem to behave properly when jumping around or falling off something.

https://user-images.githubusercontent.com/50862920/180096548-f842f397-1435-4e8d-81b4-aeff18829b89.mp4

aMannus commented:
> Ice trap tokens doesn't seem to behave properly when jumping around or falling off something.
> 
>  2022-07-21.00-56-15.mp4

I'm not entirely sure if it would matter, but was this done before or after my last commit? It did change some stuff regarding ice traps.

Flax4 commented:
Tried latest build, still happens

aMannus commented:
> Tried latest build, still happens

Thanks for checking. I'll take a look soon.

leggettc18 commented:
If it helps, the 00f8 textbox is the one in the vanilla game at that textId (hence why I usually put "unused" in quotes when I say its unused) so somehow the check that is supposed to escape into the custom text isn't happening. The falling off one where it did show the textbox is probably just something about the player state that somehow isn't caught in the current check for being in the air, if I had to guess.

aMannus committed:
Merge branch 'rando-next' into spoopy-spiders

aMannus committed:
Fixed ice traps (maybe actually this time)

aMannus commented:
> Ice trap tokens doesn't seem to behave properly when jumping around or falling off something.
> 
>  2022-07-21.00-56-15.mp4

Should be fixed now. Maybe actually this time. Also tested out the ice trap overwriting other items like the Goron Ruby from your last stream, and that was indeed already fixed with the commit I made during your run.

briaguya0 commented on soh/include/z64player.h:
```diff
@@ -5,6 +5,16 @@
+    /* 0x04 */ u16 objectId;
+} GetItemEntry; // size = 0x06
+
+extern GetItemEntry sGetItemTable[160];
```
i'm not the biggest fan of externing this, but i think a better solution is outside the scope of this PR, and shouldn't keep it from being merged into `rando-next`

briaguya0 commented on soh/include/z64player.h:
```diff
@@ -612,6 +622,7 @@ typedef struct Player {
     /* 0x0A86 */ s8         unk_A86;
     /* 0x0A87 */ u8         unk_A87;
     /* 0x0A88 */ Vec3f      unk_A88; // previous body part 0 position
+    /* 0x0A89 */ u8         iceTrapped;
```
i'm not the biggest fan of this name, maybe `pendingIceTrap`? the name should describe the state of having picked up an ice trap item but not having been frozen yet.

briaguya0 replied:
also might make sense to just make it a `bool` and use true/false instead of 0/1

aMannus replied:
Agree on both fronts, will change this.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3301,7 +3521,8 @@ void DrawRandoEditor(bool& open) {
     const char* randoLinksPocket[4] = { "Dungeon Reward", "Advancement", "Anything", "Nothing" };
-    const char* randoShuffleSongs[3] = { "Song locations", "Dungeon rewards", "Anywhere" };
+    const char* randoShuffleSongs[3] = { "Song Locations", "Dungeon Rewards", "Anywhere" };
+    const char* randoShuffleTokens[4] = { "Off", "Dungeons", "Overworld", "All Tokens" };
```
nitpick, do we need to have "All *Tokens*" here, or could it just be "All"?

aMannus replied:
I think it just adds a tiny bit more clarity. N64 and 3DS rando also both call it specifically "All tokens". The option name isn't super long as it is either, so I don't think there's any harm in keeping it this way.

briaguya0 replied:
looking at it in the menu i agree all tokens is more clear, if the option was titled "Shuffle Skulltula Tokens" having "All Tokens" would feel redundant, but since it just says "Tokensanity" it feels less redundant

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3702,6 +3923,22 @@ void DrawRandoEditor(bool& open) {
+                            "\n"
+                            "Overworld - Only shuffle GS tokens that are outside of dungeons.\n"
+                            "\n"
+                            "All Tokens - Shuffle all 100 GS tokens."
```
same as previous question about all tokens vs just all

briaguya0 reviewed (changes requested):
couple of little things, overall it's looking good!

aMannus committed:
Changed iceTrapped var name/type and tooltip

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -10978,7 +10978,7 @@ void Player_UpdateCommon(Player* this, GlobalContext* globalCtx, Input* input) {
     Collider_ResetQuadAT(globalCtx, &this->shieldQuad.base);
 
-    if (this->iceTrapped != 0) {
+    if (this->pendingIceTrap == true) {
```
no need to have `== true` here, just `if (this->pendingIceTrap)` works

aMannus replied:
Woops.

aMannus committed:
Tiny code cleanup
